org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java

```diff
@@ -13,36 +13,94 @@
 
 package org.eclipse.jdt.internal.javac;
 
+import java.io.IOException;
 import java.util.Locale;
 
 import javax.tools.Diagnostic;
 import javax.tools.JavaFileObject;
 
+import org.eclipse.core.runtime.ILog;
 import org.eclipse.jdt.core.compiler.IProblem;
 import org.eclipse.jdt.internal.compiler.problem.ProblemSeverities;
 
 import com.sun.tools.javac.code.Kinds;
 import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.util.DiagnosticSource;
 import com.sun.tools.javac.util.JCDiagnostic;
+import com.sun.tools.javac.util.Position;
 
 public class JavacProblemConverter {
 	public static JavacProblem createJavacProblem(Diagnostic<? extends JavaFileObject> diagnostic) {
+		int problemId = toProblemId(diagnostic);
+		return switch (problemId) {
+			case IProblem.MissingSerialVersion -> missingSerialVersion(diagnostic, problemId);
+			default -> defaultProblem(diagnostic, problemId);
+		};
+	}
+
+	private static JavacProblem defaultProblem(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
+		return new JavacProblem(
+				diagnostic.getSource().getName().toCharArray(),
+				diagnostic.getMessage(Locale.getDefault()),
+				diagnostic.getCode(),
+				problemId,
+				new String[0],
+				toSeverity(diagnostic),
+				(int) Math.min(diagnostic.getPosition(), diagnostic.getStartPosition()),
+				(int) (diagnostic.getEndPosition() - 1),
+				(int) diagnostic.getLineNumber(),
+				(int) diagnostic.getColumnNumber());
+	}
+
+	private static int toSeverity(Diagnostic<? extends JavaFileObject> diagnostic) {
+		return switch (diagnostic.getKind()) {
+			case ERROR -> ProblemSeverities.Error;
+			case WARNING, MANDATORY_WARNING -> ProblemSeverities.Warning;
+			case NOTE -> ProblemSeverities.Info;
+			default -> ProblemSeverities.Error;
+		};
+	}
+
+	private static JavacProblem missingSerialVersion(Diagnostic<? extends JavaFileObject> diagnostic, int problemId) {
+		int startPosition = - 1;
+		int endPosition = - 1;
+		if (diagnostic instanceof JCDiagnostic jcDiagnostic
+				&& jcDiagnostic.getDiagnosticPosition() instanceof JCClassDecl jcClassDecl) {
+			startPosition = (int) diagnostic.getPosition();
+			if (startPosition != Position.NOPOS) {
+				DiagnosticSource source = jcDiagnostic.getDiagnosticSource();
+				JavaFileObject fileObject = source.getFile();
+				try {
+					CharSequence charContent = fileObject.getCharContent(true);
+					String content = charContent.toString();
+					String name = jcClassDecl.getSimpleName().toString();
```

Comment:
Is the range of the name available as a child node of the jcClassDecl? I think it should be: https://github.com/openjdk/jdk/blob/33e81229bd1b4b28cf2e35f0f8f0a42a04d59c3d/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java#L778

Reply:
I have used it - https://github.com/openjdk/jdk/blob/33e81229bd1b4b28cf2e35f0f8f0a42a04d59c3d/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java#L829

Reply:
My bad, it seems like the Name is not an ast node, your implementation is good